compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/ExternalDrag.desktop.kt

```diff
@@ -0,0 +1,462 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.Immutable
+import androidx.compose.runtime.State
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberUpdatedState
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.AwtWindowDragTargetListener.WindowDragValue
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.geometry.Rect
+import androidx.compose.ui.graphics.painter.Painter
+import androidx.compose.ui.layout.boundsInWindow
+import androidx.compose.ui.layout.onGloballyPositioned
+import androidx.compose.ui.window.LocalWindow
+import androidx.compose.ui.window.density
+import java.awt.Point
+import java.awt.Window
+import java.awt.dnd.DnDConstants
+import java.awt.dnd.DropTarget
+import java.awt.dnd.DropTargetDragEvent
+import java.awt.dnd.DropTargetDropEvent
+import java.awt.dnd.DropTargetEvent
+import java.awt.dnd.DropTargetListener
+
+/**
+ * Represent data that is being dragged (or dropped) to a component from outside an application.
+ */
+@ExperimentalComposeUiApi
+interface DragData {
+    /**
+     * Represents list of files drag and dropped to a component.
+     */
+    interface FilesList : DragData {
+        /**
+         * Returns list of file paths drag and droppped to an application in a URI format.
+         */
+        fun readFiles(): List<String>
+    }
+
+    /**
+     * Represents an image drag and dropped to a component.
+     */
+    interface Image : DragData {
+        /**
+         * Returns an image drag and dropped to an application as a [Painter] type.
+         */
+        fun readImage(): Painter
+    }
+
+    /**
+     * Represent text drag and dropped to a component.
+     */
+    interface Text : DragData {
+        /**
+         * Provides the best MIME type that describes text returned in [readText]
+         */
+        val bestMimeType: String
+
+        /**
+         * Returns a text dropped to an application.
+         */
+        fun readText(): String
+    }
+}
+
+/**
+ * Represent the current state of drag and drop to a component from outside an application.
+ * This state is passed to external drag callbacks.
+ *
+ * @see onExternalDrag
+ */
+@ExperimentalComposeUiApi
+@Immutable
+class ExternalDragValue internal constructor(
```

Comment:
```suggestion
class ExternalDragValue(
```

If the thing is simple (as here), better to make it public. That makes Compose more easy to extend. If this case - without exposing the constructor, we can't write a wrapper around `onExternalDrag`.

The general rule is to expose everything that is well designed and minimal for maximum use cases.